### PR TITLE
VFS file watcher based use cases implementations

### PR DESCRIPTION
**Pull request by dkulieshov**
### What does this PR do?

Adds implementations for several user scenarios covering: 
- File opened in editor VFS synchronization
- Project tree synchronization
- Git checkout operations notification
### What issues does this PR fix or reference?
#1910, #1911, #1912, #1830, #1832 as part of #1824
### PR type
- [x] Major change = changes existing features
### Major change checklist
- [x] Unit tests provided/updated
- [x] Unit tests passed
- [x] Acceptance tests passed

This PR is based on squashed commits, to see the full commit history please refer to [che#1912](https://github.com/eclipse/che/tree/che%231912) branch

### CODE REVIEW

File: plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/refactoring/rename/wizard/RenamePresenter.java [hunk 1]
```diff
@@ -373,6 +383,14 @@ public void apply(RefactoringResult arg) throws OperationException {
...
                     view.showErrorMessage(arg);
                 }
+
+                for (ChangeInfo change : arg.getChanges()) {
+                    final String path = change.getPath();
+                    final String oldPath = change.getOldPath();
+
+                    eventBus.fireEvent(new FileTrackingEvent(path, oldPath, MOVE));
+                }
+                eventBus.fireEvent(new FileTrackingEvent(null, null, RESUME));
```
Comment (vparfonov) [inline plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/refactoring/rename/wizard/RenamePresenter.java]: Strange event, no? (newPath ==  null oldPath == null)

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/EditorFileStatusNotificationReceiver.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import static org.eclipse.che.ide.api.notification.StatusNotification.DisplayMode.EMERGE_MODE;
+import static org.eclipse.che.ide.api.notification.StatusNotification.Status.SUCCESS;
+import static org.eclipse.che.ide.api.resources.ResourceDelta.REMOVED;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/EditorFileStatusNotificationReceiver.java:41]: javadoc

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/FileOpenCloseEventListener.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.List;
+import java.util.Objects;
+
+import static org.eclipse.che.api.project.shared.dto.event.FileTrackingOperationDto.Type.START;
+import static org.eclipse.che.api.project.shared.dto.event.FileTrackingOperationDto.Type.STOP;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/FileOpenCloseEventListener.java:37]: javadoc

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/FileTrackingEvent.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ *******************************************************************************/
+package org.eclipse.che.ide.api.event.ng;
+
+import com.google.gwt.event.shared.EventHandler;
+import com.google.gwt.event.shared.GwtEvent;
+
+import org.eclipse.che.api.project.shared.dto.event.FileTrackingOperationDto;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/FileTrackingEvent.java:28]: javadoc

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/GitCheckoutStatusNotificationReceiver.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import org.eclipse.che.ide.util.loging.Log;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import static org.eclipse.che.ide.api.notification.StatusNotification.DisplayMode.EMERGE_MODE;
+import static org.eclipse.che.ide.api.notification.StatusNotification.Status.SUCCESS;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/GitCheckoutStatusNotificationReceiver.java]: javadoc

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/ProjectTreeStatusNotificationReceiver.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import static org.eclipse.che.ide.api.resources.ResourceDelta.ADDED;
+import static org.eclipse.che.ide.api.resources.ResourceDelta.REMOVED;
+import static org.eclipse.che.ide.api.resources.ResourceDelta.UPDATED;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/ProjectTreeStatusNotificationReceiver.java:37]: javadoc

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/EditorFileStatusDetector.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import org.eclipse.che.api.vfs.impl.file.event.EventTreeNode;
+import org.eclipse.che.api.vfs.impl.file.event.HiEvent;
+import org.eclipse.che.api.vfs.impl.file.event.HiEventDetector;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Optional;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/EditorFileStatusDetector.java]: javadoc

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingOperationReceiver.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import org.eclipse.che.dto.server.DtoFactory;
+import org.slf4j.Logger;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingOperationReceiver.java:40]: javadoc

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingOperationTransmitter.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import static org.eclipse.che.api.project.shared.dto.event.FileWatcherEventType.DELETED;
+import static org.eclipse.che.api.project.shared.dto.event.FileWatcherEventType.MODIFIED;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingOperationTransmitter.java:36]: javadoc

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingRegistry.java [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.nio.charset.Charset.defaultCharset;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingRegistry.java:40]: javadoc for class and all methods

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingRegistry.java [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+
+        if (fileTrackingMetadata == null) {
+            registry.put(path, new FileTrackingMetadata(path, endpoint));
+        } else {
+            fileTrackingMetadata.addEndpoint(endpoint);
+        }
+    }
+
+    public void suspend(int endpoint) {
+        registry.values().stream().filter(m -> m.getNotSuspendedEndpoints().contains(endpoint)).forEach(m -> m.suspend(endpoint));
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingRegistry.java]: line too long

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingRegistry.java [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    public void resume(int endpoint) {
+        registry.values().stream().filter(m -> m.getSuspendedEndpoints().contains(endpoint)).forEach(m -> m.resume(endpoint));
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingRegistry.java]: line too long

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toSet;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * @author Dmitry Kuleshov
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java:36]: javadoc

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+    public void suspend() {
+        this.state = State.SUSPENDED;
+    }
+
+    public void resume() {
+        // TODO this workaround is needed to compensate delays
+        // in file watcher system, should be removed
+        // after VFS is fixed properly
+        try {
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java:65]: link to the issue

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+            for (EventTreeNode node : directories) {
+                final String path = node.getPath();
+                final FileWatcherEventType lastEventType = node.getLastEventType();
+
+                transmit(path, lastEventType);
+            }
+        }
+
+        return Optional.empty();
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java:102]: is that ok that you returning only `return Optional.empty()` from this method

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+ * @author Dmitry Kuleshov
+ */
+@Beta
+@Singleton
+public class ProjectTreeChangesDetector implements HiEventDetector<ProjectTreeChangesDetector> {
+    private static final Logger LOG = getLogger(ProjectTreeChangesDetector.class);
+
+    private final JsonRpcRequestTransmitter transmitter;
+
+    private State state;
```
Comment (skabashnyuk) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java:45]: looks like boolean variable.

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/EditorFileStatusNotificationReceiver.java [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+import org.eclipse.che.ide.util.loging.Log;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import static org.eclipse.che.ide.api.notification.StatusNotification.DisplayMode.EMERGE_MODE;
+import static org.eclipse.che.ide.api.notification.StatusNotification.Status.SUCCESS;
+import static org.eclipse.che.ide.api.resources.ResourceDelta.REMOVED;
+
+/**
```
Comment (vparfonov) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/EditorFileStatusNotificationReceiver.java:35]: javadoc?

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/FileTrackingEvent.java [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+ */
+public class FileTrackingEvent extends GwtEvent<FileTrackingEvent.FileTrackingEventHandler> {
```
Comment (vparfonov) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/FileTrackingEvent.java:30]: javadoc

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/GitCheckoutStatusNotificationReceiver.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+ */
+@Singleton
```
Comment (vparfonov) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/GitCheckoutStatusNotificationReceiver.java]: move to git plugin

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/ProjectTreeStatusNotificationReceiver.java (hunk 6 above)
Comment (vparfonov) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/ProjectTreeStatusNotificationReceiver.java:37]: javadoc

File: ide/che-core-ide-api/src/test/java/org/eclipse/che/ide/api/event/ng/GitCheckoutStatusNotificationReceiverTest.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link GitCheckoutStatusNotificationReceiver}
+ *
+ * @author Dmitry Kuleshov
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class GitCheckoutStatusNotificationReceiverTest {
```
Comment (vparfonov) [inline ide/che-core-ide-api/src/test/java/org/eclipse/che/ide/api/event/ng/GitCheckoutStatusNotificationReceiverTest.java]: move to the git plugin

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingOperationTransmitter.java [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+ */
+@Singleton
+public class FileTrackingOperationTransmitter {
```
Comment (vparfonov) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingOperationTransmitter.java:39]: javadoc

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingRegistry.java [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.nio.charset.Charset.defaultCharset;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * @author Dmitry Kuleshov
+ */
+@Singleton
```
Comment (vparfonov) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/FileTrackingRegistry.java:42]: Javadoc

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/GitCheckoutHiEventDetector.java [hunk 23]
```diff
@@ -40,79 +42,84 @@
...
+ * the project root folder.
  * </p>
  *
  * @author Dmitry Kuleshov
- *
  * @since 4.5
  */
 @Beta
-public class GitCheckoutHiEventDetector implements HiEventDetector<GitBranchCheckoutEventDto> {
+public class GitCheckoutHiEventDetector implements HiEventDetector<GitCheckoutEventDto> {
```
Comment (vparfonov) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/GitCheckoutHiEventDetector.java:61]: move to the git plugin

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+import static java.util.stream.Collectors.toSet;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * @author Dmitry Kuleshov
+ */
+@Beta
+@Singleton
+public class ProjectTreeChangesDetector implements HiEventDetector<ProjectTreeChangesDetector> {
```
Comment (vparfonov) [inline wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java:40]: javadoc

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/ProjectTreeStatusNotificationReceiver.java (hunk 6 above)
Comment (dkulieshov) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/ProjectTreeStatusNotificationReceiver.java:37]: javadoc

Comment (svor) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/ProjectTreeStatusNotificationReceiver.java:37]: javadoc

File: plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/refactoring/rename/wizard/RenamePresenter.java (hunk 1 above)
Comment (dkulieshov) [reply plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/refactoring/rename/wizard/RenamePresenter.java]: Yes

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/ProjectTreeStatusNotificationReceiver.java (hunk 6 above)
Comment (skabashnyuk) [reply ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/event/ng/ProjectTreeStatusNotificationReceiver.java:37]: Are you sure, maybe JavaDoc?

File: wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java (hunk 16 above)
Comment (dkulieshov) [reply wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/vfs/impl/file/event/detectors/ProjectTreeChangesDetector.java:45]: Enum is used to make it more declarative, in general - looks like boolean can be used here.

**Comment by vparfonov** [comment]
ok

**Pull request merged**